File: app/src/commands/cheat/behavior.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { invoke } from '@tauri-apps/api/core';
import type { BuildStep } from '@/types/core/behavior';
import { CoordImpl } from '@/core/model/continent/coord';
import type { ContinentKey } from '@/types/core/continent';
import type { CheatGetBuildStepsRequest } from '@/types/request/cheat/behavior';

export async function cheatGetBuildSteps(
  coord: ContinentKey,
  limit = Number.MAX_SAFE_INTEGER,
) {
  coord = CoordImpl.fromContinentKey(coord);
  const req: CheatGetBuildStepsRequest = {
    world: NIL.world.getIdStrict(),
    coord,
  };

  const steps = await invoke<BuildStep[]>('cheat_get_build_steps', { req });
  while (steps.length > limit) steps.pop();
  return steps as readonly BuildStep[];
}

File: crates/nil-server/src/router/cheat/behavior.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::res;
use crate::response::EitherExt;
use crate::state::App;
use axum::extract::{Json, State};
use axum::response::Response;
use nil_payload::cheat::behavior::*;

pub async fn get_build_steps(
  State(app): State<App>,
  Json(req): Json<CheatGetBuildStepsRequest>,
) -> Response {
  app
    .world_mut(req.world, |world| world.cheat_get_build_steps(req.coord))
    .await
    .try_map_left(|steps| res!(OK, Json(steps)))
    .into_inner()
}
